### ISSUE TITLE
关于您的工作是否可迁移的疑问

### ISSUE THREAD

**Issue by qd121** [liwk19/NORA#3]
您好，打扰一下，我有几点疑问想向您咨询。
我的任务是想要通过评估在一个因果图上去除某个节点（节点表示事件，边表示因果关系（且边只有这一种关系）），即模拟反事实，来得到去除该事件节点对于其他节点的影响。请问这样的任务是否可直接应用您的代码逻辑？
期待您的回复！！

**Comment by weikai-li** [comment]
感谢您对我们工作的关注！我们的方法基于graph neural network (GNN)，如果您也使用GNN的话，我们的代码是可以用的。我们的assumption是GNN的prediction可以模拟反事实的情况。我们的setting是在original graph上训练好GNN模型，之后删除某个节点，再在新graph上跑一遍GNN inference，我们希望GNN在新graph上的prediction可以模拟反事实的label。我们提出的方法可以快速估算整个graph上所有节点被删除时的影响。我们的方法主要适用的场景是：（1）这个graph适合使用GNN，GNN的prediction比较准，这样我们才能信任GNN的prediction去模拟反事实的情况；（2）您的graph比较大，或者需要估算的node比较多，需要快速估算的方法来节省时间，否则可以直接去除node，然后再跑一遍GNN inference也很快

如果还有什么问题欢迎继续留言~

**Comment by qd121** [comment]
感谢您的回复，再次打扰，我看到您的边任务中，    nora的边任务需要① 节点→边：先从节点嵌入构造边得分
② 边→节点：再把边得分聚合回节点，                                                       
     我的模型本身通过CGE和mlp,得到了prediction即边的分数，需要像nora代码中一样再聚集回节点吗？还是这个并不是硬性的，另外，这种情况是否是否可直接应用您的代码逻辑？

**Comment by weikai-li** [comment]
感谢您对我们工作的持续关注！我没有太理解 “① 节点→边：先从节点嵌入构造边得分 ② 边→节点：再把边得分聚合回节点” 这句话。我们在做link prediction task的时候，选择了一种常见的方式，通过两个node的node embedding去predict link，这里就是“① 节点→边”。您说的 “② 边→节点” 是指哪一步呢？是指估算link prediction任务里node influence的时候吗？

**Comment by qd121** [comment]
感谢您的回复，我指的是您的代码中的这步，                                                                                                                                                    ①out = out[pred_edge_ori[0]] * out[pred_edge_ori[1]]  # 
            out = out.sum(1)
            out = torch.sigmoid(out)    #  边的概率                                                                                                                                         以及 ②link_info1 = torch.concat([pred_edge_ori[0], pred_edge_ori[1]])
            link_info2 = torch.concat([pred_edge_ori[1], pred_edge_ori[0]])
            tmp_g = dgl.graph((link_info1, link_info2), num_nodes=num_node).to(device)
            tmp_g.edata['score'] = torch.concat([out, out])
            tmp_g.update_all(fn.copy_e('score', 'score_m'), fn.sum('score_m', 'score_sum'))
            out = tmp_g.ndata['score_sum']  # 每个节点的邻居边分数和。                                                                                                                   这里我的理解是                                                                                                                                                                                                          ①通过两个节点的节点嵌入去预测链接得到了  out = torch.sigmoid(out)    #  边的概率                                                                      ② 最后这里的out = tmp_g.ndata['score_sum']是把边得分聚合回节点，即得到每个节点的分数 ，也就是我上述提到的 边→节点  。                                                                                                                                                                                       那在我的代码中，前提是：有图和节点表示，要通过图注意力卷积得到更新后的节点表示event_embed后，                                                                 然后 得到  event_diff = event_embed[[event_pairs[0][0]]] - event_embed[[event_pairs[0][1]]]
            event_pair_pre = torch.cat((event_pair_embed, event_diff), dim=1) 
            prediction = self.mlp(event_pair_pre)  #即边分数的预测，                                                                                                                              然后，我该怎么迁移使用您的方法来计算节点影响力呢？                                                                                                                                是要像②一样把边得分聚合回节点吗？                                                                                                                                                                     期待您的回复，万分感谢！

**Comment by weikai-li** [comment]
谢谢您的解释！我现在明白您的意思了。您的理解基本是正确的，我们在做link prediction task的时候，采用的方法是concatenate两个node embedding，然后根据concatenated embedding去做预测。

我们在估算node influence score的时候，对于node classification task，我们会使用node classification prediction（paper里的f_r）加上一些基于gradient和heuristics的估算方式，去估算node influence score。对于link prediction task，我们把link prediction score聚合回node，之后的步骤就和node classification里估算node influence score一样了。分散回node的这个link prediction score本身并不是我们估算的node influence score，我们会使用它以及一些gradient和heuristics的方式，去得到对node influence score的估算。

另外提醒您需要注意的一点是，您需要通过”直接去除node，然后在新graph上跑一遍GNN inference“的方式获得一部分node的ground-truth influence score。然后根据这些node的ground-truth influence score去调整我们用NORA方法估算node influence score时使用的超参数，这样估算才能比较准确。

**Comment by qd121** [comment]
非常感谢您的回复！！

**Comment by weikai-li** [comment]
没事没事。那我关掉这个issue了，如果后续还有问题欢迎继续提问
